oculus_ros2/include/oculus_ros2/sonar_viewer.hpp

```diff
@@ -68,70 +70,65 @@ std::vector<double> linspace(T start_in, T end_in, int num_in) {
   }
 
   double delta = (end - start) / (num - 1);
-
-  for (int i = 0; i < num - 1; ++i) {
-    linspaced.push_back(start + delta * i);
-  }
+  std::generate(linspaced.begin(), linspaced.end(), [i = 0, &delta]() mutable { return i++ * delta; });
   linspaced.push_back(end);  // I want to ensure that start and end
                              // are exactly the same as the input
   return linspaced;
 }
 
-template <typename dataType>
-void SonarViewer::publish_fan(const int& width,
+template <typename DataType>
+void SonarViewer::publishFan(const int& width,
     const int& height,
     const int& offset,
-    const std::vector<dataType>& ping_data,
+    const std::vector<DataType>& ping_data,
     const int& master_mode,
     const double& ping_range,
     const std_msgs::msg::Header& header) const {
+  static_assert(std::is_same<DataType, uint8_t>::value || std::is_same<DataType, uint16_t>::value,
+      "publishFan can only be build for uint8_t and uint16_t");
+
+  const int step = width + SIZE_OF_GAIN;
+
   // Create rawDataMat from ping_data
-  cv::Mat rawDataMat(height, (width + 4), CV_8U);
-  std::memcpy(rawDataMat.data, ping_data.data() + offset, height * (width + 4));
+  cv::Mat rawDataMat(height, step, CV_8U);
+  std::memcpy(rawDataMat.data, ping_data.data() + offset, height * step);
 
-  int bearing = (master_mode == 1) ? 65 : 40;
-  std::vector<double> ranges = linspace(0., ping_range, height);
-  int image_width = 2 * std::sin(bearing * M_PI / 180) * ranges.size();
+  const double bearing = (master_mode == 1) ? LOW_FREQUENCY_BEARING_APERTURE : HIGHT_FREQUENCY_BEARING_APERTURE;
+  const std::vector<double> ranges = linspace(0., ping_range, height);
+  const int image_width = 2 * std::sin(bearing * M_PI / 180) * ranges.size();
   cv::Mat mono_img;
-  if (std::is_same<dataType, uint8_t>::value) {
-    mono_img = cv::Mat::ones(cv::Size(image_width, ranges.size()), CV_8UC1) * 255;
-  } else {
-    mono_img = cv::Mat::ones(cv::Size(image_width, ranges.size()), CV_16UC1) * 65535;
-  }
-
-  // for (int i = 0; i < image_width; i++) {
-  //   for (int j = 0; j < ranges.size(); j++) {
-  //     mono_img.at<dataType>(j, i) = 255;
-  //   }
-  // }
+  const int cv_encoding = std::is_same<DataType, uint8_t>::value ? CV_8UC1 : CV_16UC1;
```

Comment:
This is the same as previously but shorter : the check will be performed each time you execute the function.
What you can do is
```cpp
if constexpr (std::is_same<DataType, uint8_t>::value) {
  mono_img = cv::Mat::ones(cv::Size(image_width, ranges.size()), CV_8UC1) *
             std::numeric_limits<DataType>::max();
} else {
  mono_img = cv::Mat::ones(cv::Size(image_width, ranges.size()), CV_16UC1) *
             std::numeric_limits<DataType>::max();
}
```

The compiler will check at build time the condition and only implement the correct version.